docs/_posts/2021-07-05-whats-next.md

```diff
@@ -0,0 +1,176 @@
+---
+layout: post
+title: "⏭️ After the browser-state migration - what's next?"
+date: 2021-07-05 14:00:00 +0200
+author: sebastian
+---
+
+After 2+ years slowly and incrementally working towards this goal, we [completed the migration](https://github.com/mozilla-mobile/android-components/pull/10436) from the `browser-session` component to the `browser-state` component for state handling. Finally, we were able to delete `browser-session` and all state is now maintained and updated by the [redux](https://redux.js.org/introduction/core-concepts)-like [`BrowserStore`](https://github.com/mozilla-mobile/android-components/blob/master/components/browser/state/src/main/java/mozilla/components/browser/state/store/BrowserStore.kt#L22).
+
+The following blog posting describes some of the possible follow-up changes to the architecture that we would consider, depending on the outcome of further discussions and prototyping.
+
+## Reversing the dependency between browser-state and browser-engine implementations
+
+In the current architecture every `browser-engine` implements `concept-engine` and exposes an abstracted mechanism for observing events. For every `EngineSession` an `EngineObserver` gets created, which will dispatch a `BrowserAction` for every event, updating the centralized state.
+
+![](/assets/images/blog/engine-architecture.png)
+
+Now that the migration to `browser-state` is completed, we can reverse this dependency. With a `browser-engine` implementation depending on `browser-state` directly, it can dispatch actions without an observer in between.
+
+![](/assets/images/blog/simplified-engine-architecture.png)
+
+This removes the requirement for a shared, abstracted observer interface in `concept-engine`. It would no longer be required to have shared _“glue code”_ for connecting an abstract browser engine with the state handling component.
+
+A potential downside to this approach is that the `BrowserStore` and related `BrowserAction`s become the new _“interface”_ that a browser engine has to dispatch correctly, which can be harder to understand and follow than simply implementing actual interfaces.
+
+Overall this seems to be worthwhile exploring and potentially discussing further in an RFC.
```

Comment:
Among other things, this would also greatly simplify our web extension implementation. In addition to dropping `EngineObserver`, we could also remove `WebExtensionDelegate` and the glue logic in between (the `support-webextension` module).

However, I am going back and forth on this. From a consumer point of view, having an explicit interface is definitely easier to understand and deal with than a loosely specified set of actions they can observe, even if documented perfectly.

I am really curious what everyone thinks about this? Perfect case for an RFC and prototype :).

Reply:
Yeah, fully agree. I really like the simplification that comes from it, but it may make things harder to grasp.